### PR TITLE
Make `length` member read-only

### PR DESCRIPTION
**Pull request by erak**
Fixes https://github.com/ethereum/solidity/issues/3515.

- [x] Fix solc-js test (Blocked by https://github.com/ethereum/solidity/pull/7475 and https://github.com/ethereum/solc-js/pull/386)
- [x] Documentation
- [x] Reactivate compilation via Yul where applicable
- [x] Changelog entry
- [x] Fix spellcheck: rebase after develop has merged to develop_060

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp
Comment (leonardoalt) [inline libsolidity/analysis/TypeChecker.cpp]: Can be removed.

File: libsolidity/analysis/TypeChecker.cpp
Comment (leonardoalt) [inline libsolidity/analysis/TypeChecker.cpp:2401]: Could assert inside the if that the member's name is `length`.

File: test/libsolidity/semanticTests/viaYul/array_storage_index_boundery_test.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/array_storage_index_boundery_test.sol]: Indentation

File: test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol]: Same here and below

File: test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol]: And here

**Comment by erak** [comment]
Tests are mainly failing because some of them can't be compiled via yul (`push()` and `pop()` are not implemented yet). Created an issue to track that: https://github.com/ethereum/solidity/issues/7351

File: libsolidity/analysis/TypeChecker.cpp
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:2873]: Can we add a meaningful error for all cases?

File: test/libsolidity/semanticTests/storage/array_accessor.sol
Comment (chriseth) [inline test/libsolidity/semanticTests/storage/array_accessor.sol]: maybe `smallTypeData = new uint24[](128);`?

File: libsolidity/analysis/TypeChecker.cpp
Comment (ekpyron) [reply libsolidity/analysis/TypeChecker.cpp:2401]: Wouldn't that potentially break e.g. ``using SomeLibrary for uint256[];`` and then ``uint256[] a; a.someLibraryFunction()``?

File: test/libsolidity/syntaxTests/array/calldata_resize.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/array/calldata_resize.sol:7]: Why do you kill the error messages I just added recently and delete their tests ;-)?

File: libsolidity/analysis/TypeChecker.cpp
Comment (erak) [reply libsolidity/analysis/TypeChecker.cpp:2873]: Yes, that makes sense and is also in line with @ekpyron's comment below :)

File: libsolidity/analysis/TypeChecker.cpp
Comment (erak) [reply libsolidity/analysis/TypeChecker.cpp:2401]: I'll add a test for that.

**Comment by erak** [comment]
There's one test left that fails due to a broken stack after accessing `storageC.ma.length`:
```
contract Test {
    struct A {
        mapping(uint=>uint) m;
    }
    struct B {
        mapping(uint=>uint) m;
        uint x;
    }
    struct C {
        mapping(uint=>uint)[] ma;
    }
    A storageA;
    B storageB;
    C storageC;
    constructor() public {
        storageA.m[1] = 2;
        storageB.m[3] = 4;
        storageB.x = 5;
    }
    function g() public returns (uint, uint) {
        A memory memoryA = A();
        B memory memoryB = B(42);
        C memory memoryC = C();
        storageA = memoryA;
        storageB = memoryB;
        storageC = memoryC;
        return (storageA.m[1], storageC.ma.length);
    }
```

File: test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol
Comment (erak) [inline test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol:17]: Here it would have been nice to be able to run a test suite with different evm versions :)

File: Changelog.md
Comment (chriseth) [inline Changelog.md]: "always read-only, even for storage arrays.

File: docs/060-breaking-changes.rst
Comment (chriseth) [inline docs/060-breaking-changes.rst]: or assign a full array, which will of course overwrite existing content.

Comment (chriseth) [reply docs/060-breaking-changes.rst]: or use pop

File: docs/060-breaking-changes.rst
Comment (chriseth) [inline docs/060-breaking-changes.rst]:
```suggestion
* Change ``array.length++`` to ``array.push()`` to increase, and use ``pop()`` to decrease
```

Comment (chriseth) [reply docs/060-breaking-changes.rst]: please mention both `array.push()` and `array.push(value)`.

File: docs/types/reference-types.rst
Comment (chriseth) [inline docs/types/reference-types.rst:288]: But now, also existing pairs will be cleared.

**Comment by erak** [comment]
Updated.

File: docs/types/reference-types.rst
Comment (leonardoalt) [inline docs/types/reference-types.rst]:
```suggestion
                m_byteData.push();
```

File: test/compilationTests/corion/schelling.sol
Comment (leonardoalt) [inline test/compilationTests/corion/schelling.sol]:
```suggestion
        rounds.push();
```

File: test/compilationTests/milestonetracker/MilestoneTracker.sol
Comment (leonardoalt) [inline test/compilationTests/milestonetracker/MilestoneTracker.sol]:
```suggestion
            Milestone storage milestone = milestones.push();
```
magic

File: test/contracts/Wallet.cpp
Comment (leonardoalt) [inline test/contracts/Wallet.cpp:211]:
```suggestion
			m_pendingIndex.push(_operation);
			pending.index = m_pendingIndex.length - 1;
```

File: test/libsolidity/SolidityOptimizer.cpp
Comment (leonardoalt) [inline test/libsolidity/SolidityOptimizer.cpp]:
```suggestion
					data1.push();
```

File: test/libsolidity/SolidityOptimizer.cpp
Comment (leonardoalt) [inline test/libsolidity/SolidityOptimizer.cpp]:
```suggestion
				container.valueIndices.push();
```

File: test/libsolidity/semanticTests/storage/array_accessor.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/storage/array_accessor.sol]:
```suggestion
        dynamicData.push();
```

File: test/libsolidity/semanticTests/structs/conversion/recursive_storage_memory.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/structs/conversion/recursive_storage_memory.sol]: Won't this push the same reference all over?

File: test/libsolidity/semanticTests/structs/conversion/recursive_storage_memory_complex.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/structs/conversion/recursive_storage_memory_complex.sol:25]: Spaces instead of tabs from here on

File: test/libsolidity/semanticTests/viaYul/array_storage_index_access.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/array_storage_index_access.sol]:
```suggestion
    function test_indices(uint256 len) public
```

File: test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol:2]: Spaces instead of tabs

**Review by leonardoalt — COMMENTED** [review]
Can you also add tests that try to resize the length of an array that
- is the field of a storage struct
- is the value of a mapping

File: libsolidity/analysis/TypeChecker.cpp
Comment (erak) [reply libsolidity/analysis/TypeChecker.cpp:2401]: @ekpyron So, is this the use case you had in mind? It still compiles :-)
```
library Numbers {
    function process(uint256[] storage self) external {}
}
contract TestUnit {
    using Numbers for uint256[];
    uint256[] a;
    function f() public {
        a.process();
    }
}
```

File: test/compilationTests/milestonetracker/MilestoneTracker.sol
Comment (erak) [reply test/compilationTests/milestonetracker/MilestoneTracker.sol]: :-)

File: test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol
Comment (erak) [reply test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol:2]: Mmh, to me they now look correct. In the majority of syntax and semantic tests we do use spaces. This is a first unification.

Comment (erak) [reply test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol]: Same as in my comment above applies.

File: test/libsolidity/semanticTests/structs/conversion/recursive_storage_memory_complex.sol
Comment (erak) [reply test/libsolidity/semanticTests/structs/conversion/recursive_storage_memory_complex.sol:25]: Same as in my other comment applies.

File: test/libsolidity/semanticTests/structs/conversion/recursive_storage_memory.sol
Comment (erak) [reply test/libsolidity/semanticTests/structs/conversion/recursive_storage_memory.sol]: Ah, true. This is a left over and I now changed it to empty `push()`.

**Comment by erak** [comment]
Updated.

File: libsolidity/analysis/TypeChecker.cpp
Comment (ekpyron) [reply libsolidity/analysis/TypeChecker.cpp:2401]: Yes, I meant asserting that the name is "length", as @leonardoalt suggested, might have broken that, so better not to assert that.

File: test/libsolidity/syntaxTests/viewPureChecker/access_to_base_member_array.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/viewPureChecker/access_to_base_member_array.sol:7]: Is there a different test for ``push()`` not to be ``view``?

File: test/libsolidity/syntaxTests/nameAndTypeResolution/219_memory_arrays_not_resizeable.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/nameAndTypeResolution/219_memory_arrays_not_resizeable.sol:8]: Is there a test for ``x.push()`` to fail for memory arrays?

File: test/libsolidity/syntaxTests/array/calldata_resize.sol
Comment (ekpyron) [inline test/libsolidity/syntaxTests/array/calldata_resize.sol:7]: Should there be a test for ``x.push()`` to fail for calldata arrays?

File: test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol
Comment (ekpyron) [inline test/libsolidity/semanticTests/viaYul/array_storage_index_zeroed_test.sol:45]: Would it hurt to implement ``push`` for Yul codegen here as well? It looks like using ``m_utils.resizeDynamicArrayFunction`` in the ``MemberAccess`` visiter there or something like that should just work.

**Review by ekpyron — COMMENTED** [review]
Generally looks good to me - but do we want Yul IR codegen here as well? And I'm a bit worried about the test coverage - I'd expect a new test using ``push`` for each removed test with an assignment to ``length``...

**Comment by Marenz** [comment]
needs rebasing due to conflicts

**Comment by ekpyron** [comment]
I'll rebase this now and will take a look into implementing ``push`` in Yul codegen. @erak Ping me when you're back on, so that we can coordinate!

**Comment by ekpyron** [comment]
Status for this: I rebased it, fixed/adjusted stuff in the docs that was added and am now starting to implement push and pop for yul codegen, but that won't really touch the rest of this PR for now.

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `acebf8faa1b389f67b9bb02f1382476ddcce1c96`:
```
Coding style error:
 libsolidity/codegen/YulUtilFunctions.cpp:596:std::string YulUtilFunctions::storageArrayPopFunction(const dev::solidity::ArrayType& _type)
 libsolidity/codegen/YulUtilFunctions.cpp:626:std::string YulUtilFunctions::storageArrayPushFunction(const dev::solidity::ArrayType& _type)

```
Please check that your changes are working as intended.

**Comment by ekpyron** [comment]
I now implemented some basic storage array ``push`` and ``pop``, replaced the ``= new uint[](...)`` assignments to storage arrays by using push and pop instead (otherwise we'd need array copying from memory to storage...) and restored all disabled "viaYul" tests that way. Maybe some more dedicated tests for push and pop via yul are in order, though...

**Comment by erak** [comment]
@ekpyron I've added some more tests. Would you mind checking again?

**Comment by ekpyron** [comment]
@erak The test failures are probably rebase issues? Apart from that maybe someone else could/should review this, since I wrote parts of it myself and I'll be off for a while now?

File: docs/types/reference-types.rst
Comment (leonardoalt) [inline docs/types/reference-types.rst]:
```suggestion
    Increasing the length of a storage array by calling ``push()``
```

File: docs/types/reference-types.rst
Comment (leonardoalt) [inline docs/types/reference-types.rst]:
```suggestion
    while decreasing the length by calling ``pop()`` has at least
```

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
	solUnimplementedAssert(_type.baseType()->storageBytes() <= 32, "");
```

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.cpp]: This doesn't seem to be used

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (leonardoalt) [inline libsolidity/codegen/YulUtilFunctions.cpp]:
```suggestion
	solUnimplementedAssert(_type.baseType()->storageBytes() <= 32, "");
```

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (leonardoalt) [inline test/libsolidity/SolidityEndToEndTest.cpp]: Is there an issue for this already? It would maybe be a good idea to open one before this gets merged otherwise we might forget to re-enable it.

File: test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol]: Why `petersburg`?

File: test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol
Comment (leonardoalt) [inline test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol:18]: Why does it run out of gas?

**Review by leonardoalt — COMMENTED** [review]
A few more small comments.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (erak) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I've also cleaned up this and the line below.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (erak) [reply libsolidity/codegen/YulUtilFunctions.cpp]: I've also cleaned up this and the line below.

File: libsolidity/codegen/YulUtilFunctions.cpp
Comment (erak) [reply libsolidity/codegen/YulUtilFunctions.cpp]: :+1:

File: test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol
Comment (erak) [reply test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol]: Removed it.

File: test/libsolidity/SolidityEndToEndTest.cpp
Comment (erak) [reply test/libsolidity/SolidityEndToEndTest.cpp]: Good catch. This doesn't need to be deactivated any more, since `push()` via yul is available now.

Comment (erak) [reply test/libsolidity/SolidityEndToEndTest.cpp]: There are some more test cases which I re-enabled already.

Comment (erak) [reply test/libsolidity/SolidityEndToEndTest.cpp]: @leonardoalt The remaining ones are tracked with https://github.com/ethereum/solidity/issues/7749.

**Comment by leonardoalt** [comment]
Needs rebase

File: test/libsolidity/syntaxTests/viewPureChecker/access_to_base_member_array.sol
Comment (erak) [reply test/libsolidity/syntaxTests/viewPureChecker/access_to_base_member_array.sol:7]: Added.

File: test/libsolidity/syntaxTests/nameAndTypeResolution/219_memory_arrays_not_resizeable.sol
Comment (erak) [reply test/libsolidity/syntaxTests/nameAndTypeResolution/219_memory_arrays_not_resizeable.sol:8]: Now there is :-)

File: test/libsolidity/syntaxTests/array/calldata_resize.sol
Comment (erak) [reply test/libsolidity/syntaxTests/array/calldata_resize.sol:7]: Now there is :-)

File: test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol
Comment (erak) [reply test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol:18]: The yul code generated for `push()` seems to have higher gas costs than writing to `length`. Trying to narrow that down...

Comment (chriseth) [reply test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol:18]: I would rather assume it's an infinite loop...

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: No type conversion?

Comment (chriseth) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Actually I think, we should use `expressionAsType` instead of `m_context.variable`, unless this is about an lvalue or an "arbitrary type" thing.

Comment (erak) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: @chriseth Changed it such that `expressionAsType` is being used instead.

File: test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol
Comment (leonardoalt) [reply test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol:18]: @erak Any updates here?

Comment (erak) [reply test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol:18]: @leonardoalt Yes. We're simply running out of gas because the loop is of course more expensive than a simple assignment. The same happens independently from `push` by just incrementing a counter of type `uint` in the body of the loop.

The above can be run ~9500 times on emvc with our current EVMHost settings before there is no gas left any more. I think we can keep the changes to the expectations, can't we?

Comment (leonardoalt) [reply test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol:18]: Sure, I think that's fine. Just good to know the reason

**Review by leonardoalt — APPROVED** [review]

File: test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol
Comment (chriseth) [reply test/libsolidity/semanticTests/viaYul/array_storage_length_access.sol:18]: Oh I'm sorry! I totally read this the wrong way. This is exactly the behaviour we wanted to enforce by this change...

**Pull request merged**
